react-components/src/data-providers/core-dm-provider/CoreDm3dDataProvider.ts

```diff
@@ -0,0 +1,210 @@
+  readonly _fdmSdk: FdmSDK;
+
+  readonly _relevant3dSpaces: DmsUniqueIdentifier[];
+
+  readonly _revisionIdToDMSIdentifier = new Map<number, DmsUniqueIdentifier>();
+  readonly _modelIdToDMSIdentifier = new Map<number, DmsUniqueIdentifier[]>();
+
+  constructor(relevant3dSpaces: DmsUniqueIdentifier[], fdmSdk: FdmSDK) {
+    this._fdmSdk = fdmSdk;
+    this._relevant3dSpaces = relevant3dSpaces;
+  }
+
+  is3dView(view: ViewItem): boolean {
+    return view.implements.some(
+      (type) =>
+        type.externalId === COGNITE_3D_OBJECT_SOURCE.externalId &&
+        type.space === COGNITE_3D_OBJECT_SOURCE.space &&
+        type.version === COGNITE_3D_OBJECT_SOURCE.version
+    );
+  }
+
+  async getDMSModels(modelId: number): Promise<DmsUniqueIdentifier[]> {
+    const cachedModels = this._modelIdToDMSIdentifier.get(modelId);
+    if (cachedModels !== undefined) {
+      return cachedModels;
+    }
+
+    const models = await getDMSModels(modelId, this._fdmSdk);
+
+    this._modelIdToDMSIdentifier.set(modelId, models);
+    return models;
+  }
+
+  private async getDMSRevision(
+    modelRef: DmsUniqueIdentifier,
+    revisionId: number
+  ): Promise<DmsUniqueIdentifier> {
+    let revisionRef = this._revisionIdToDMSIdentifier.get(revisionId);
+
+    if (revisionRef === undefined) {
+      const revisionResult = await getDMSRevision(modelRef, revisionId, this._fdmSdk);
+
+      this._revisionIdToDMSIdentifier.set(revisionId, revisionResult);
+      revisionRef = revisionResult;
+    }
+
+    if (revisionRef === undefined) {
+      throw Error(`No revision with id ${revisionId} found`);
+    }
+
+    return restrictToDmsId(revisionRef);
+  }
+
+  private async getDMSModelsForIds(
+    modelIds: number[]
+  ): Promise<Array<DmsUniqueIdentifier | undefined>> {
+    return (
+      await executeParallel(
+        modelIds.map((id) => async () => await this.getDMSModels(id)),
+        MAX_PARALLEL_QUERIES
+      )
+    ).flat();
+  }
+
+  private async getDMSRevisionsForRevisionIdsAndModelRefs(
+    modelRefs: Array<DmsUniqueIdentifier | undefined>,
+    revisionIds: number[]
+  ): Promise<DmsUniqueIdentifier[]> {
+    return (
+      await executeParallel(
+        revisionIds.map((revisionId, ind) => async () => {
+          return modelRefs[ind] === undefined
+            ? undefined
+            : await this.getDMSRevision(modelRefs[ind], revisionId);
+        }),
+        MAX_PARALLEL_QUERIES
+      )
+    ).filter(isDefined);
+  }
+
+  async getEdgeConnected3dInstances(instance: DmsUniqueIdentifier): Promise<DmsUniqueIdentifier[]> {
+    return await getEdgeConnected3dInstances(instance, this._fdmSdk);
+  }
+
+  async getFdmConnectionsForNodes(
+    models: DmsUniqueIdentifier[],
+    revisionId: number,
+    nodes: Node3D[]
+  ): Promise<FdmCadConnection[]> {
+    if (models.length !== 1) {
+      throw new Error(`Expected 1 model, got ${models.length}`);
+    }
+
+    const model = models[0];
```

Comment:
I think this way of having an argument as an array (in this case: `models`) , suggesting multiple instances but in the function description it ends up forcing the first one to always be taken and having to be unique, is a bit contradictory in terms of semantic coding in relation to what the function requests and what it does.
Perhaps if we could make it explicit in the method name that the argument is an array with a single instance, I feel it would be less bad. But that is just my opinion and not something crucial.

Reply:
Right... This is to conform to the interface - where it could happen in the legacy FDM model, that there were multiple 3D model FDM instances for one model-ID. But I think maybe that would also count as a bug :thinking: In any case, I think I'll rather change this to a warning if there is more than one model